plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/s3/S3AsyncClientWrapper.java

```diff
@@ -0,0 +1,172 @@
+    public CompletableFuture<PutObjectResponse> putObject(PutObjectRequest request, AsyncRequestBody body)
+    {
+        CompletableFuture<PutObjectResponse> future = delegate.putObject(request, body);
+        handle(PUT_OBJECT, future);
+        return future;
+    }
+
+    @Override
+    public CompletableFuture<DeleteObjectsResponse> deleteObjects(DeleteObjectsRequest request)
+    {
+        CompletableFuture<DeleteObjectsResponse> future = delegate.deleteObjects(request);
+        handle(DELETE_OBJECTS, future);
+        return future;
+    }
+
+    @Override
+    public <T> CompletableFuture<T> getObject(GetObjectRequest request, AsyncResponseTransformer<GetObjectResponse, T> transformer)
+    {
+        CompletableFuture<T> future = delegate.getObject(request, transformer);
+        handle(GET_OBJECT, future);
+        return future;
+    }
+
+    @Override
+    public CompletableFuture<CreateMultipartUploadResponse> createMultipartUpload(CreateMultipartUploadRequest request)
+    {
+        CompletableFuture<CreateMultipartUploadResponse> future = delegate.createMultipartUpload(request);
+        handle(CREATE_MULTIPART_UPLOAD, future);
+        return future;
+    }
+
+    @Override
+    public CompletableFuture<UploadPartResponse> uploadPart(UploadPartRequest request, AsyncRequestBody body)
+    {
+        CompletableFuture<UploadPartResponse> future = delegate.uploadPart(request, body);
+        handle(UPLOAD_PART, future);
+        return future;
+    }
+
+    @Override
+    public CompletableFuture<CompleteMultipartUploadResponse> completeMultipartUpload(CompleteMultipartUploadRequest request)
+    {
+        CompletableFuture<CompleteMultipartUploadResponse> future = delegate.completeMultipartUpload(request);
+        handle(COMPLETE_MULTIPART_UPLOAD, future);
+        return future;
+    }
+
+    @Override
+    public CompletableFuture<AbortMultipartUploadResponse> abortMultipartUpload(AbortMultipartUploadRequest request)
+    {
+        CompletableFuture<AbortMultipartUploadResponse> future = delegate.abortMultipartUpload(request);
+        handle(ABORT_MULTIPART_UPLOAD, future);
+        return future;
+    }
+
+    @Override
+    public CompletableFuture<ListObjectsV2Response> listObjectsV2(ListObjectsV2Request request)
+    {
+        CompletableFuture<ListObjectsV2Response> future = delegate.listObjectsV2(request);
+        handle(LIST_OBJECTS_V2, future);
+        return future;
+    }
+
+    protected abstract void handle(RequestType requestType, CompletableFuture<?> responseFuture);
+
+    @Override
+    public ListObjectsV2Publisher listObjectsV2Paginator(ListObjectsV2Request listObjectsV2Request)
+    {
+        return new ListObjectsV2Publisher(this, applyPaginatorUserAgent(listObjectsV2Request));
+    }
+
+    /**
+     * Based on {@link software.amazon.awssdk.services.s3.DefaultS3AsyncClient#applyPaginatorUserAgent(S3Request)}
+     */
+    private <T extends S3Request> T applyPaginatorUserAgent(T request)
```

Comment:
where exactly are the separate requests for pages handled?

Reply:
This is done by passing the reference to the wrapper to the `ListObjectsV2Publisher`

> return new ListObjectsV2Publisher(this, applyPaginatorUserAgent(listObjectsV2Request));